### PR TITLE
clang-tools: move into llvmPackages

### PR DESCRIPTION
**Pull request by ShamrockLee**
###### Description of changes

Format the Nix expression of `clang-tools` using `nixfmt`, in accordance with NixOS/rfcs#166.

Move the The clang-tools expression folder under `llvm/common` and `clang-tools` package under `llvmPackages`.

Move `clang-tools_<version>` into `aliases.nix`, specifying as `llvmPackages_<version>.clang-tools`.

~Provide `clang-tools-python` that links and wraps executables from `clang-unwrapped.python` output. If applied, users will be able to enjoy `git clang-format` with `clang-tools-python` installed (or introduced inside Nix shell).~

~As for `"${clang-unwrapped.python}/bin/scan-view"`, it depends on the relative module inside `"../share"`. The dependent module of `scan-view`, `ScanView.py`, is currently not moved into `clang-unwrapped.python`, which will be fixed by the staging PR #191801.~

Cc:
@Patryk27 clang-tools maintainer

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) (The 24 updated packages will take some time to build.)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`) (Tested git-clang-format, but not scan-view)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ShamrockLee** [comment]
@ofborg build bat-extras.prettybat cloudcompare entwine grass libpulsar nominatim pdal python310Packages.tiledb python39Packages.tiledb qgis qgis-ltr tiledb vscode-extensions.ms-vscode-cpptools

**Review by Patryk27 — APPROVED** [review]

**Comment by ShamrockLee** [comment]
Just seperate those Python scripts to a new output `python`, so that packages depending on clang-tools won't be forced to have Python in their closure.

**Review by Patryk27 — APPROVED** [review]
Looks pretty neat 🙂

**Review by Patryk27 — APPROVED** [review]

**Comment by ShamrockLee** [comment]
Could it be merged now?

Or should I split the wrapped python script into a new derivation `clang-tools-python` to reduce the number of rebuild?

**Comment by Patryk27** [comment]
The changes look fine to me -- I don't personally have merging rights, though 👀

**Comment by ShamrockLee** [comment]
No wonder. I had thought that all the members have the merging right.

**Comment by ShamrockLee** [comment]
Just make a new derivation `clang-tools-python` instead. This way, packages depending on `clang-tools` won't have to rebuild whenever python gets rebuild.

Wrapping the Python scripts into a new package also guarantees zero rebuild of the current packages, and is thus easier to backport.

**Review by Patryk27 — APPROVED** [review]
🚀

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:13680]: Why don't we move it under llvmPackages?

File: pkgs/development/tools/clang-tools/python.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/clang-tools/python.nix:57]: What about homepage?

File: pkgs/development/tools/clang-tools/python.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/clang-tools/python.nix:25]: That is unused I think

File: pkgs/development/tools/clang-tools/python.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/clang-tools/python.nix:24]: and this should be inline

File: pkgs/development/tools/clang-tools/python.nix
Comment (ShamrockLee) [reply pkgs/development/tools/clang-tools/python.nix:57]: The `homepage` attribute is inherited from `clang-unwrapped.meta`.

`clang-tools` and `clang-tools-python` are merely wrappers to the corresponding files provided by the `clang-unwrapped` package, so the homepage should be the same.

Comment (ShamrockLee) [reply pkgs/development/tools/clang-tools/python.nix:57]: Update: I have changed my mind after seeing `meta.name` and `meta.mainProgram` from clang-unwrapped.meta.name`.

The attributes to inherit should be explicitly specified as you suggested for the `raspa` PR (https://github.com/NixOS/nixpkgs/pull/186127#discussion_r1039016406).

File: pkgs/top-level/all-packages.nix
Comment (ShamrockLee) [reply pkgs/top-level/all-packages.nix:13680]: Sounds great.

Not sure what would be the proper way to move it into `llvmPackages`. It's quite confusing to see such one-version-per-folder form of packaging.

Comment (ShamrockLee) [reply pkgs/top-level/all-packages.nix]: Just moved `clang-tools` and `clang-tools-python` into `llvmPackages`.

**Comment by ShamrockLee** [comment]
@ofborg build clang-tools-python clang-tools-python.passthru.tests

File: pkgs/top-level/all-packages.nix
Comment (ShamrockLee) [reply pkgs/top-level/all-packages.nix]: Update: Strange errors occurred.

**Comment by ShamrockLee** [comment]
@SuperSandro2000 I don't know why, but it keeps complaining

```
error: Function called without required argument "clang-tools-extra_src" at /run/media/root/data-btrfs/shamrock-shared/targets/Projects/NixOS/nixpkgs_optparse/pkgs/development/compilers/llvm/5/clang/default.nix:1
```
and
```
error: Function called without required argument "llvm_meta" at /run/media/root/data-btrfs/shamrock-shared/targets/Projects/NixOS/nixpkgs_optparse/pkgs/development/compilers/llvm/14/clang/default.nix:1
```

I just specify
```
clang-tools = callPackage ../common/clang-tools {
  inherit (tools) clang-unwrapped;
};
```
as other packages in llvmPackages_5 does.

**Comment by ShamrockLee** [comment]
Issues solved!

File: pkgs/top-level/all-packages.nix
Comment (ShamrockLee) [reply pkgs/top-level/all-packages.nix]: Addressed.

**Comment by ShamrockLee** [comment]
Do we consider

1. Changing the `<pkg>.override` interface
2. Moving packages into `aliases.nix`

breaking changes?

**Comment by ShamrockLee** [comment]
Refine the package test.

**Comment by ShamrockLee** [comment]
Bump onto master and resolve merge conflict.

**Comment by ShamrockLee** [comment]
I'm splitting this PR to preserve only the `clang-tools`'s movig to `llvmPackages` part. This is a backward-incompatible change, and I hope it could go into Nixpkgs 24.05.

This also allows me to re-think if `clang-tools-python` is the suitable way to provide `clang-tools-python`. I just noticed the existence of `clang-analyzer`, which provides `scan-build` and `scan-view`. After removing `scan-view` from `clang-tools-python`, all it provides will be `git-clang-format`.

**Comment by Patryk27** [comment]
Hi, sure! I'll take a look tomorrow 🙂

**Review by Patryk27 — APPROVED** [review]

**Comment by ShamrockLee** [comment]
I bumped it onto the `master` branch, resolved merge conflicts, and revised the release note to use imperative tones.

**Review by Patryk27 — APPROVED** [review]

**Review by RossComputerGuy — APPROVED** [review]
Just resolve the conflicts and this should be good to go.

**Comment by ShamrockLee** [comment]
@RossComputerGuy The release note entry is inside `rl-2405.section.md`. Is it okay to backport this PR, or how should I handle the release note entry?

**Comment by RossComputerGuy** [comment]
@ShamrockLee I don't think it's too late to backport.

**Comment by rrbutani** [comment]
> Is it okay to backport this PR

> I don't think it's too late to backport

I think we should maybe get a second opinion on this; breaking changes are restricted <4 weeks before [24.05's release](https://github.com/NixOS/nixpkgs/issues/303285 and changing the `clang-tools_*` attributes to aliases could be construed as breaking.

> Imo, release note entries are not that important for something like this.

I think it's worth having, especially since the intent (as per the OP) is to phase out the aliases that were added.

**Comment by ShamrockLee** [comment]
@rrbutani Not sure if the change of `<pkg>.override` interfaces caused by moving `clang-tools` into `llvmPakcages` would also be considered backward-incompatible.

**Review by rrbutani — CHANGES_REQUESTED** [review]
Looks good, couple of nits about the release note entry:

File: pkgs/development/compilers/llvm/common/clang-tools/default.nix
Comment (rrbutani) [inline pkgs/development/compilers/llvm/common/clang-tools/default.nix:17]: nit: the scope that's used in `llvmPackages` already has `version`:

https://github.com/NixOS/nixpkgs/blob/0179f8c547879eb26016fb1dcda2904001487a3d/pkgs/development/compilers/llvm/18/default.nix#L61

You should be able to drop `clang-unwrapped` from this package's attrs (and all of the `llvmPackages_*` set's `default.nix`s and just ask for `version` in this package's attrs.

**Review by rrbutani — CHANGES_REQUESTED** [review]

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (ShamrockLee) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Addressed. Thank you for pointing out.

By the way, I once received a review from @fricklerhandwerk in another PR that suggests using imperative tones when writing release notes. However, most of the entries in this edited release note use passive tones so that the target of changes becomes the first word of a sentence.

Do you think I should change the newly added entry and use a passive tone instead?

**Comment by rrbutani** [comment]
> Not sure if the change of `<pkg>.override` interfaces caused by moving `clang-tools` into `llvmPakcages` would also be considered backward-incompatible.

oh, good point! here, let's ask:

---

@wegank This PR changes some top-level attrs to be aliases and changes some package attributes such that existing `clang-tools_<ver>.override { ... }` invocations may break (now takes `clang`, `libcxxClang` instead of `llvmPackages`). We're wondering if this is okay to backport to 24.05 or if this is considered a breaking change.

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (rrbutani) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: As far as I know indicative mood with the subject up-front is the aim here but @fricklerhandwerk likely knows better than I do; I don't interact with the release notes much 🙂 

cc: @jacg

**Comment by wegank** [comment]
I'd consider the change of argument for `clang-tools` to be breaking. There's also the move to aliases for `clang-tools_1{2-8}`, which sounds breaking too.

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (jacg) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: TLDR: yes, please change the entry to match the style used throughout the rest of document.

For the 24.05 release, I was instructed to ensure that the name of the affected item appear at or near the beginning of every note (in order to make it easier to spot relevant entries when scanning the document), hence the subject-up-front indicative style is used throughout the document. This was already the prevailing style before I started editing the notes.

I would therefore argue that, in this context, all notes should follow that style, regardless of what merits other styles might have in other contexts.

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (ShamrockLee) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Addressed.

File: pkgs/development/compilers/llvm/common/clang-tools/default.nix
Comment (ShamrockLee) [reply pkgs/development/compilers/llvm/common/clang-tools/default.nix:17]: Since `clang-tools` is a wrapper of (a part of) `clang-unwrapped`, the build process would be more straightforward to pass `clang-unwrapped` as a shell variable. It's also okay to string-interpolate it directly.

As for the version, it makes sense to pass it from the llvmPackage an `<pkg>.override` argument.

> You should be able to drop clang-unwrapped from this package's attrs

Are we removing the attribute to prevent people from accessing the `clang-unwrapped` package used via the attribute (or even try to override it via `<pkg>.overrideAttrs`) but encourage accessing/overriding via `llvmPackages` instead?

Comment (rrbutani) [reply pkgs/development/compilers/llvm/common/clang-tools/default.nix:17]: My apologies, I missed that the `$out/bin` symlinks are produced from `clang-unwrapped` and not `clang`/`libcxxClang`.

Please disregard the above; getting `version` from `clang-unwrapped` seems like the sensible thing to do.

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (rrbutani) [inline nixos/doc/manual/release-notes/rl-2405.section.md:240]: Since we're not backporting this PR we should move this to the [24.11 release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md#backward-incompatibilities-sec-release-2411-incompatibilities).

**Review by rrbutani — CHANGES_REQUESTED** [review]
@ShamrockLee thanks for working on this; one last minor change and this will be good to merge:

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (ShamrockLee) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Addressed.

**Review by rrbutani — APPROVED** [review]

**Comment by ShamrockLee** [comment]
I just rebased the feature branch on top of the `master` branch and resolved the merge conflict of the release note entries.

Is there anything I could help to push it forward?

**Comment by ShamrockLee** [comment]
I simplified the package calling of `clang-tools` in each `llvmPackages` and fixed the evaluation, i.e.

```diff
     # Wrapper for standalone command line utilities
-    clang-tools = callPackage ../common/clang-tools {
-      inherit (tools) clang-unwrapped clang libcxxClang;
-      inherit llvm_meta;
-    };
+    clang-tools = callPackage ../common/clang-tools { };
```

**Review by Patryk27 — APPROVED** [review]

**Comment by JohnRTitor** [comment]
Hi, could you rebase and remove the conflicts please? I'll merge it straight after that.

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 191698` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>llvmPackages_12.clang-tools</li>
    <li>llvmPackages_13.clang-tools</li>
    <li>llvmPackages_14.clang-tools</li>
    <li>llvmPackages_15.clang-tools</li>
    <li>llvmPackages_16.clang-tools</li>
    <li>llvmPackages_18.clang-tools</li>
  </ul>
</details>

**Review by JohnRTitor — APPROVED** [review]

**Pull request merged**
